sol/main.c

```diff
@@ -0,0 +1,513 @@
+	}
+
+	req_queue->len++;
+}
+
+#define GK_REQ_ENQ_BURST_SIZE (32)
+#define GK_REQ_DEQ_BURST_SIZE (32)
+
+static void
+enqueue_reqs(struct sol_config *sol_conf)
+{
+	struct priority_req *req_nodes[GK_REQ_ENQ_BURST_SIZE];
+	int num_reqs = mb_dequeue_burst(&sol_conf->mb,
+		(void **)req_nodes, GK_REQ_ENQ_BURST_SIZE);
+	int i;
+	for (i = 0; i < num_reqs; i++)
+		req_enqueue(sol_conf, req_nodes[i]);
+}
+
+/* Scaling for cycles_per_byte calculation
+ * Chosen so that minimum rate is 480 bit/sec
+ */
+#define RTE_SCHED_TIME_SHIFT		      8
+
+static inline void
+time_resync(struct req_queue *req_queue)
+{
+	uint64_t cycles = rte_get_tsc_cycles();
+	uint64_t cycles_diff = cycles - req_queue->time_cpu_cycles;
+	uint64_t bytes_diff;
+
+	/* Compute elapsed time in bytes. */
+	bytes_diff = rte_reciprocal_divide(cycles_diff << RTE_SCHED_TIME_SHIFT,
+		req_queue->inv_cycles_per_byte);
+
+	/* Advance port time. */
+	req_queue->time_cpu_cycles = cycles;
+	req_queue->time_cpu_bytes += bytes_diff;
+	if (req_queue->time < req_queue->time_cpu_bytes)
+		req_queue->time = req_queue->time_cpu_bytes;
+}
+
+static inline uint32_t
+min_val_2_u32(uint32_t x, uint32_t y)
+{
+	return (x < y) ? x : y;
+}
+
+static inline void
+credits_update(struct req_queue *req_queue)
+{
+	uint64_t n_periods = (req_queue->time - req_queue->tb_time) /
+		req_queue->tb_period;
+	req_queue->tb_credits += n_periods * req_queue->tb_credits_per_period;
+	req_queue->tb_credits = min_val_2_u32(req_queue->tb_credits,
+		req_queue->tb_size);
+	req_queue->tb_time += n_periods * req_queue->tb_period;
+}
+
+static inline int
+credits_check(struct req_queue *req_queue, struct rte_mbuf *pkt)
+{
+	uint32_t pkt_len = pkt->pkt_len + req_queue->frame_overhead;
+
+	if (pkt_len > req_queue->tb_credits)
+		return 0;
+
+	req_queue->tb_credits -= pkt_len;
+	return 1;
+}
+
+static void
+dequeue_reqs(struct sol_config *sol_conf, uint8_t tx_port,
+	uint16_t tx_queue)
+{
+	struct req_queue *req_queue = sol_conf->req_queue;
+	struct priority_req *head = req_queue->head;
+
+	struct rte_mbuf *pkts_out[GK_REQ_DEQ_BURST_SIZE];
+	uint32_t nb_pkts_out = 0;
+	uint16_t sent = 0;
+
+	time_resync(req_queue);
+
+	/*
+	 * XXX When is the best time to do this? DPDK sched does
+	 * this for every packet that it sees while doing prefetching.
+	 */
+	credits_update(req_queue);
+
+	while (nb_pkts_out < GK_REQ_DEQ_BURST_SIZE && head != NULL) {
+
+		/* Get packet out of linked list. */
+		struct rte_mbuf *pkt = head->pkt;
+
+		/* Check credits. */
+		if (!credits_check(req_queue, pkt)) {
+			printf("out of credits\n");
+			goto out;
+		}
+
+		/* Remove request from queue. */
+		if (head->prev == NULL ||
+			(head->priority != head->prev->priority)) {
+			req_queue->priorities[head->priority] = NULL;
+			rte_bitmap_clear(req_queue->bmp, head->priority);
+		}
+		head = head->prev;
+		if (head != NULL)
+			head->next = NULL;
+		req_queue->len--;
+
+		/* Remove extra space in mbuf. */
+		if (rte_pktmbuf_adj(pkt, sizeof(*head)) == NULL) {
```

Comment:
Why does one need to remove ``sizeof(*head)`` space in ``mbuf``? It seems that ``mbuf`` is the packet received from NIC without any change.